File: src/main/java/net/zomis/tttultimate/TTWinCondition.java
package net.zomis.tttultimate;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;

public class TTWinCondition implements Iterable<Winnable> {

	private final List<Winnable> winnables;
	private final int consecutive;

	public TTWinCondition(Winnable... winnables) {
		this(Arrays.asList(winnables));
	}
	
	public TTWinCondition(List<? extends Winnable> winnables) {
		this(winnables, winnables.size());
	}
	
	public TTWinCondition(List<? extends Winnable> winnables, int consecutive) {
		if (winnables.isEmpty())
			throw new IllegalArgumentException("Can't have an empty win condition!");
		this.winnables = Collections.unmodifiableList(new ArrayList<Winnable>(winnables));
		this.consecutive = consecutive;
	}
	
	public int neededForWin(TTPlayer player) {
		return winnables.size() - hasCurrently(player);
	}
	public boolean isWinnable(TTPlayer byPlayer) {
		return hasCurrently(byPlayer.next()) == 0;
	}
	public int hasCurrently(TTPlayer player) {
		int i = 0;
		for (Winnable winnable : winnables) {
			if (winnable.getWonBy().and(player) == player)
				i++;
		}
		return i;
	}
	
	public TTPlayer determineWinnerNew() {
		TTPlayer winner = TTPlayer.NONE;
		
		int[] consecutivePlayers = new int[TTPlayer.values().length];
		for (Winnable winnable : winnables) {
			TTPlayer current = winnable.getWonBy();
			for (TTPlayer pl : TTPlayer.values()) {
				int i = pl.ordinal();
				if (pl.and(current) == pl) {
					consecutivePlayers[i]++;
				}
				else consecutivePlayers[i] = 0;
				
				if (consecutivePlayers[i] >= this.consecutive) {
					winner = winner.or(pl);
				}
			}
		}
		return winner;
	}
	
	public boolean hasWinnable(Winnable field) {
		return winnables.contains(field);
	}
	
	public int size() {
		return winnables.size();
	}

	@Override
	public Iterator<Winnable> iterator() {
		return new ArrayList<Winnable>(this.winnables).iterator();
	}
	
}

File: uttt-common/src/main/kotlin/net/zomis/tttultimate/TTWinCondition.kt
package net.zomis.tttultimate

class TTWinCondition : Iterable<TTBase> {

    private val winnables: List<TTBase>
    private val consecutive: Int

    constructor(vararg winnables: TTBase) : this(winnables.toList())

    constructor(winnables: List<TTBase>) : this(winnables, winnables.size)

    constructor(winnables: List<TTBase>, consecutive: Int) {
        if (winnables.isEmpty()) {
            throw IllegalArgumentException("Can't have an empty win condition!")
        }
        this.winnables = winnables.toMutableList()
        this.consecutive = consecutive
    }

    fun neededForWin(player: TTPlayer): Int {
        return winnables.size - hasCurrently(player)
    }

    fun isWinnable(byPlayer: TTPlayer): Boolean {
        return hasCurrently(byPlayer.next()) == 0
    }

    fun hasCurrently(player: TTPlayer): Int {
        var i = 0
        for (winnable in winnables) {
            if (winnable.wonBy.and(player) == player) {
                i++
            }
        }
        return i
    }

    fun determineWinnerNew(): TTPlayer {
        var winner: TTPlayer = TTPlayer.NONE

        val consecutivePlayers = IntArray(TTPlayer.values().size)
        for (winnable in winnables) {
            val current = winnable.wonBy
            for (pl in TTPlayer.values()) {
                val i = pl.ordinal
                if (pl.and(current) == pl) {
                    consecutivePlayers[i]++
                } else {
                    consecutivePlayers[i] = 0
                }

                if (consecutivePlayers[i] >= this.consecutive) {
                    winner = winner.or(pl)
                }
            }
        }
        return winner
    }

    fun hasWinnable(field: TTBase): Boolean {
        return winnables.contains(field)
    }

    fun size(): Int {
        return winnables.size
    }

    override fun iterator(): Iterator<TTBase> {
        return this.winnables.toMutableList().iterator()
    }

}
